promql/prettier/prettier.go

```diff
@@ -0,0 +1,499 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prettier
+
+import (
+	"regexp"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+const (
+	// itemNotFound is used to initialize the index of an item.
+	itemNotFound = -1
+)
+
+// Prettier handles the prettifying and formatting operation over a
+// list of rules files or a single expression.
+type Prettier struct {
+	expression string
+	Node       parser.Expr
+	pd         *padder
+}
+
+type padder struct {
+	indent      string
+	columnLimit int
+	// isNewLineApplied represents a new line was added previously, after
+	// appending the item to the result. It should be set to default
+	// after applying baseIndent. This property is dependent across the padder states
+	// and hence needs to be carried along while formatting.
+	isNewLineApplied bool
+	// isPreviousItemComment confirms whether the previous formatted item was
+	// a comment. This property is dependent on previous padder states and hence needs
+	// to be carried along while formatting.
+	isPreviousItemComment bool
+	// expectAggregationLabels expects the following item(s) to be labels of aggregation
+	// expression. Labels are used in AggregateExpr and in VectorSelector but only the ones
+	// in VectorSelector are splittable.
+	expectAggregationLabels bool
+	// labelsSplittable determines whether the current formatting region lies in labels
+	// matchers in VectorSelector. Since labels in VectorSelectors are only splittable,
+	// is property is applied only after encountering a Left_Brace. This is dependent across
+	// multiple padder states and hence needs to be carried along.
+	labelsSplittable bool
+}
+
+// New returns a new prettier over the given slice of files.
+func New(expression string) *Prettier {
+	return &Prettier{
+		expression: expression,
+		pd:         &padder{indent: "  ", isNewLineApplied: true, columnLimit: 100},
```

Comment:
If we're hardcoding then these may be better as file-level consts. That'l help avoid deviation by different users over time.

Presume that everything here is going to end up public, as that is the policy.

Reply:
Addressed this as well. Please comment on its satisfactoriness.